### PR TITLE
SNOW-1763555 update how HTTP headers are requested from OpenTelemetry

### LINKED ISSUE #2084: SNOW-1763555: OpenTelemetry integration utterly breaks driver if using DataDog
### Python version

N/A

### Operating system and processor architecture

N/A

### Installed packages

```python
ddtrace
opentelemetry-api
```

### What did you do?

I used the driver w/ Datadog APM, which installs the opentelemetry packages that are used as the check to see if we can inject trace context.

With DataDog, this gets happens on every query, and it fails to run.

This is because the only exception caught is ModuleNotFoundError, and opentracing is throwing a ValueError which isn't getting caught.

(please forgive any typos, I copied the backtrace from a screenshot)

```python
ValueError: Propagator tracecontext not found. It is either misspelled or not installed.
  File "/var/task/lambda_function.py", line 116, in lambda_handler
    process_scan(
  File "/var/task/store.py", line 92, in process_scan
    upload_file_to_snowflake(cursor, secrets, file_name=sv_file, table=snowflake_table)
  File "/var/task/store.py", line 55, in upload_file_to_snowflake
    cursor.execute (
  File "/var/task/snowflake/connector/cursor.py", line 984, in execute
    ret = self._execute_helper(query, **kwargs)
  File "/var/task/snowflake/connector/cursor.py", line 695, in _execute_helper
    ret = self._connection.cmd_query(
  File "/var/task/snowflake/connector/connection.py", line 1350, in cd_query
    ret = self. rest. request(
  File "/var/task/snowflake/connector/network.py", line 483, in request
    from opentelemetry.propagate import inject
  File "<frozen importlib._bootstrap»", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap»", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap»", Line 690, in _load_unlocked
  File "/var/task/ddtrace/internal/module.py", line 295, in _exec_module
    self.loader.exec_module (module)
  File "/var/task/opentelemetry/propagate/__init__.py", line 149, in ‹module>
    raise ValueError(
```

### What did you expect to see?

I would expect that opentelemetry integration would be done via this- https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation

and not via a hack in the actual library that breaks the driver for anyone that has opentelemetry installed but not in-use.

I hope that this issue results in that change being reverted, and the original requestor of the feature gets this done properly in the open-telemetry project.

### Can you set logging to DEBUG and collect the logs?

```bash
We already have it at DEBUG and that's the problem.
```

### PR DESCRIPTION
**Pull request by sfc-gh-mkeller**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #2084 

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   In this PR I fix an error where sometime injecting OpenTelemetry headers into our own causes Exceptions to be thrown.

4. (Optional) PR for stored-proc connector:

### CODE REVIEW

File: src/snowflake/connector/network.py [hunk 1]
```diff
@@ -481,11 +481,17 @@ def request(
             HTTP_HEADER_USER_AGENT: PYTHON_CONNECTOR_USER_AGENT,
         }
         try:
-            from opentelemetry.propagate import inject
+            # SNOW-1763555: inject OpenTelemetry headers if available specifically in WC3 format
+            #  into our request headers in case tracing is enabled. This should make sure that
+            #  our requests are accounted for properly if OpenTelemetry is used by users.
+            from opentelemetry.propagate import TraceContextTextMapPropagator
```
Comment (sfc-gh-bdrutu) [inline src/snowflake/connector/network.py:487]: from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

File: src/snowflake/connector/network.py [hunk 2]
```diff
@@ -481,11 +481,17 @@ def request(
...
 
-            inject(headers)
-        except ModuleNotFoundError as e:
-            logger.debug(f"Opentelemtry otel injection failed because of: {e}")
+            TraceContextTextMapPropagator().inject(headers)
+        except Exception:
+            logger.debug(
+                "Opentelemtry otel injection failed",
+                exc_info=True,
+            )
```
Comment (sfc-gh-bdrutu) [inline src/snowflake/connector/network.py:494]: Do we have a test about this log message? It happened in the past that we crash because of this.

**Review by sfc-gh-bdrutu — APPROVED** [review]

File: src/snowflake/connector/network.py (hunk 1 above)
Comment (sfc-gh-bdrutu) [inline src/snowflake/connector/network.py:487]:
```suggestion
            from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
```

File: src/snowflake/connector/network.py (hunk 2 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/network.py:494]: I can add one, not a problem!

**Review by sfc-gh-enie — APPROVED** [review]

**Pull request merged**